### PR TITLE
Gale and Stef -- Carets

### PR DESCRIPTION
**Pull request by galestorm**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | It would be strange to make an instance of scoring, but we still want to access the Scrabble scoring behavior. With a class method, we don't have to make a scoring object to use the scoring data and methods.   |
|   Describe an example of a test you wrote for a _nominal case_. |  When we tested to make sure that instances of player were initialized with a name parameter, we made sure that when we make a player and pass "Gale" as the argument, player.name is "Gale".	|
|   Describe an example of a test you wrote for an _edge case_. |  Since Scrabble words can be a maximum of of 7 letters, we wrote a test to make sure that argument errors would be raised if there was an attempt to play or score a string longer than 7 characters. 	|
|   What was your final test coverage percentage?	|  97.47% 	|
|   Is there a specific piece of code you'd like feedback on?	|  Player.rb lines 23-37 - we wrote a question in the comments. We're also curious re: how to elegantly make sure that the player has the right number of tiles to play a word with repeated letters (e.g. does the player have two :A tiles to play Ada).	|

### CODE REVIEW

**Comment by tildeee** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  x  |
|   Both Teammates contributed to the codebase	|  uncertain via commits  |
|   Regular Commits with meaningful commit messages  |  x  |
|  **Scoring class**	|   	|
|  `score` method	|  x  |
|   Uses appropriate data structure to store the letter score	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  x  |
|  `highest_scoring_word` method	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|  excellent!  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  x  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  x  |
|  Returns `true` or `false` from `won?` method |  x  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  x  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  x  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  x  |
| `tiles_remaining` method returns a count	|  x  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  x  |

File: lib/player.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      return @total_score
+    end
+
+    def play(word)
+      raise ArgumentError.new("Word must be a string") if word.class != String
+      raise ArgumentError.new("Word must be 7 letters or less") if word.length > 7
+      # We abandoned trying to verify that a player has the needed tiles because
+      # it looks like we would have to adjust a bunch of our tests.
+      # Question: is it normal/acceptable to go back and change tests?
+      # Or is that indicative of serious design problems?
```
Comment (tildeee) [inline lib/player.rb:28]: As a rule: It's okay to adjust tests retroactively! Does this indicate a design problem? Totally depends on the context!
I'm a little curious about why you'd have to adjust tests to make that new feature work.

File: lib/scoring.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        end
+      else # In case of ties:
+        high_score = scores.max
+        array.reject! { |word| score(word) < high_score } # Remove lower scoring words
+        array.each do |word|
+          if word.length == 7 # First 7-letter high-scorer wins
+            return word
+          end
+        end
+        return array.min_by { |word| word.length } # If no 7-letter high-scorer, shortest high-scorer wins
```
Comment (tildeee) [inline lib/scoring.rb:40]: nifty method implementation!

**Comment by tildeee** [comment]
Clean, readable code. Good work!

I tried to answer your question in the code comments, feel free to reply to them directly on that conversation and tag me with @tildeee
